kong/api/routes/cache.lua

```diff
@@ -1,18 +1,60 @@
 local responses = require "kong.tools.responses"
 local cache = require "kong.tools.database_cache"
 
+local cache_names = {}
+-- Create a table of functions per cache type
+local caches = setmetatable({
+  lua = function(action, ...)
+    if action == "get" then
+      return cache.get(...)
+    elseif action == "delete" then
+      return cache.delete(...)
+    elseif action == "delete_all" then
+      return cache.delete_all(...)
+    end
+  end,
+  shm = function(action, ...)
+    if action == "get" then
+      return cache.sh_get(...)
+    elseif action == "delete" then
+      return cache.sh_delete(...)
+    elseif action == "delete_all" then
+      return cache.sh_delete_all(...)
+    end
+  end,
+}, {
+  __index = function(self, key)
+    return responses.send_HTTP_BAD_REQUEST("invalid cache type; '"..
+           tostring(key).."', valid caches are: "..cache_names)
+  end,
+})
+
+-- build string with valid cache names (for error mesage above)
+for name in pairs(caches) do cache_names[#cache_names+1] = "'"..name.."'" end
+table.sort(cache_names)  -- make order deterministic, for test purposes
+cache_names = table.concat(cache_names, ", ")
+
+
+local function get_cache(self)
+  self.params.cache = caches[ngx.req.get_uri_args()["cache"] or "lua"]
```

Comment:
We are storing a function inside Lapis' `self.params` table? That just doesn't seem right... Also, we are calling `ngx.req.get_uri_args()` when the content of the `cache` URI argument is already in `self.params.cache`, isn't it?

How about storing it in `ngx.ctx` or even `self.kong_cache`?